### PR TITLE
Fix dispose race condition at ZipDataCacheProvider

### PR DESCRIPTION
**Pull request by Martin-Molinero**
#### Description

- `ZipDataCacheProvider.CleanCache()` will not dispose or remove entry if its being used
   - In the edge case, an entry is disposed before securing the lock on it, the entry will be re created and added.
- Reduce `DateTime.UtcNow` usages
- Just one thread will execute `CleanCache()` at the same time

### Tests
 `BasicTemplateOptionTradesAlgorithm 2018 to 2018-6. Reducing time in the cache to 0.1 seconds to reproduce the issue more frequently`

`master`
- 7 exceptions, didn't finish running, algorithm thread died/dead locked?
- 3 exceptions
- 3 exceptions
- 9 Exceptions

`pr - no exceptions`
151.66 seconds at 880k data points per second. Processing total of 133,435,554 data points.
149.40 seconds at 893k data points per second. Processing total of 133,435,554 data points.
146.89 seconds at 908k data points per second. Processing total of 133,435,554 data points.
143.47 seconds at 930k data points per second. Processing total of 133,435,554 data points.

`pr` using 10 seconds cache timeout
78.93 seconds at 1691k data points per second. Processing total of 133,435,554 data points.
77.19 seconds at **1729k data points per second**. Processing total of 133,435,554 data points.

#### Related Issue

Closes https://github.com/QuantConnect/Lean/issues/3363
#### Motivation and Context

Make `ZipDataCacheProvider` more reliable
#### Requires Documentation Change

N/A
#### How Has This Been Tested?

Regression and unit tests
#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 1]
```diff
@@ -170,26 +160,80 @@ public void Dispose()
...
         /// </summary>
-        private void CleanCache()
+        private void CleanCache(DateTime utcNow)
         {
-            var clearCacheIfOlderThan = DateTime.UtcNow.AddSeconds(-CacheSeconds);
-
-            // clean all items that that are older than CacheSeconds than the current date
-            foreach (var zip in _zipFileCache)
+            // we just want one thread cleaning the cache at the same time
+            if (Monitor.TryEnter(_zipFileCache))
```
Comment (mchandschuh) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: Nice usage of `Monitor.TryEntry` here -- no point in everyone waiting, let the others skip 👍

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 2]
```diff
@@ -71,48 +72,37 @@ public Stream Fetch(string key)
...
                         try
                         {
                             lock (existingEntry)
                             {
-                                stream = CreateStream(existingEntry, entryName, filename);
+                                if (existingEntry.Disposed)
+                                {
+                                    // bad luck, thread race condition
+                                    // it was disposed and removed after we got it
+                                    // so lets create it again and add it
```
Comment (mchandschuh) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: thank you for the comment describing this case.

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 3]
```diff
@@ -170,26 +160,80 @@ public void Dispose()
...
             }
+        }
+
+        private Stream CacheAndCreateStream(string filename, string entryName, DateTime utcNow)
+        {
+            Stream stream = null;
+            var dataStream = _dataProvider.Fetch(filename);
 
-            _lastCacheScan = DateTime.UtcNow;
+            if (dataStream != null)
```
Comment (mchandschuh) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: Consider inverting this `if` to reduce nesting:
``` csharp
if (dataStream == null)
{
    return null;
}
```

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 4]
```diff
@@ -170,26 +160,80 @@ public void Dispose()
...
+            {
+                try
+                {
+                    var newItem = new CachedZipFile(dataStream, filename, utcNow);
+
+                    // here we don't need to lock over the cache item
+                    // because it was still not added in the cache
+                    stream = CreateStream(newItem, entryName, filename);
+
+                    if (!_zipFileCache.TryAdd(filename, newItem))
```
Comment (mchandschuh) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: `ConcurrentDictionary<TKey, TValue>` provides an `AddOrUpdate` method that may be better suited here. Another pattern I've seen in the past is to use `ConcurrentDictionary<TKey, Lazy<TValue>>` which provides thread-safety on object creation by virtue of the `Lazy<T>`. What we have here _will_ work, but we'll also be creating and disposing an entry randomly when we don't need to. Using the `Lazy<T>` will prevent this case from happening.

When using the `ConcurrentDictionary<TKey, Lazy<TValue>>` you would use
```csharp
_zipFileCache.GetOrAdd(filename, fname => new Lazy<Stream>(() => /** lazy stream creation**/)
```

and we'll be guaranteed that the returned `Lazy<Stream>` is the same one used by all thread. We'll also be guaranteed that only a _single_ stream will be created. [Here's](https://blogs.endjin.com/2015/10/using-lazy-and-concurrentdictionary-to-ensure-a-thread-safe-run-once-lazy-loaded-collection/) an article I found describing the pattern. Also of interest, is the [Remarks](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.8) section of the `ConcurrentDictionary` MSDN entry for `GetOrAdd` describing the locking methodology used which guarantees that the returned instance is the same for all thread _but does not_ guarantee that only a single delegate is executed, hence the usage of `Lazy<T>` to ensure we only execute the factory delegate once.

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 5]
```diff
@@ -170,26 +160,80 @@ public void Dispose()
...
+                        newItem.Dispose();
+                    }
+                }
+                catch (Exception exception)
+                {
+                    if (exception is ZipException || exception is ZlibException)
+                    {
+                        Log.Error("ZipDataCacheProvider.Fetch(): Corrupt zip file/entry: " + filename + "#" + entryName + " Error: " + exception);
+                    }
+                    else throw;
```
Comment (mchandschuh) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: nit - bracing style -- in general, this repository always uses braces.

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 6]
```diff
@@ -290,12 +334,18 @@ public class CachedZipFile : IDisposable
         /// </summary>
         public string Key { get; private set; }
 
+        /// <summary>
+        /// Returns if this cached zip file is disposed
+        /// </summary>
+        public bool Disposed { get; private set; }
```
Comment (mchandschuh) [inline Engine/DataFeeds/ZipDataCacheProvider.cs]: Any concerns about this backing field _not_ being marked as `volatile`? Perhaps we need a `Thread.MemoryBarrier()` each time we set or retrieve the value to prevent CPU re-ordering of the instructions? Here's a great [stackoverflow](https://stackoverflow.com/a/3556877/1582922) answer discussing the lock-release semantics of .net framework as opposed to the ECMA spec itself (which uses a weaker memory model, and IIRC, mono follows the weaker model leading to bugs). The concern is the 'freshness' of the value, in other words, if ThreadA sets `Disposed=true`, then how long before ThreadB sees that `true` value since it may have cached the value from main memory. The  memory barrier (or also `volatile`) will force the CPU to retrieve the latest value from main memory (if memory serves :) )
